### ISSUE TITLE
I managed to double the training speed of this repo / keras (improvement)

### ISSUE THREAD

**Issue by zungam** [matterport/Mask_RCNN#581]
Hi, this is not a matterport issue, actually its an improvement to Keras so I will post this on Keras soon,
but it actually doubled the training speed of this repo in particular, so I will share it! I analyzed this repo for bottlenecks, and deduced the biggest bottleneck down to the `model.data_generator` which fetch images ready on a seperate thread to improve speed. However, in practice is doesnt help much, because training on a GPU is superfast (im on 1080 ti), and all `data_generator/load_mask/load_image` with all its numpy operations are super slow in this repo. This means that even though Keras fetch an image ready for each batch, the GPU still has to wait for the next image to finish loading up. 

To make the GPU allways have an image ready to load into GPU memory without waiting time, I propose having a FIFO queue fetching images non-stop. This means that while the GPU is busy doing what it does, the CPU will always work in the background to have images ready (forever). This litle trick doubled the training time of my 704x704 images with 100 segments on each image, and if you use a fast GPU like me, I believe it will help for everyone else.

**How to:**  
In keras/engine/traning.py, add this class:
`import threading
class ThreadingBatches(object):
    """ Threading example class
    The run() method will be started and it will run in the background
    until the application exits.
    """

    def __init__(self, generator=None):
        """ Constructor
        :type interval: int
        :param interval: Check interval, in seconds
        """
        self.generator = generator
        self.stack = []
        self.queue_size = 50 #This number should be experimented with based on how much RAM you have
        thread = threading.Thread(target=self.run, args=())
        thread.daemon = True                    # Daemonize thread
        thread.start()                                  # Start the execution
        
        

    def run(self):
        """ Method that runs forever """
        while True:
            # Do something
            if(len(self.stack) < self.queue_size):
                try:
                    batch = next(self.generator)
                    self.stack.append(batch)
                except ValueError:
                    None
    def popNextBatch(self):
        while len(self.stack) == 0:
            None

        return self.stack.pop(0)`

Now, add this to line 2222:
`batch_thread = ThreadingBatches(generator=output_generator) `
and overwrite line 2231 with:
`generator_output = batch_thread.popNextBatch()`

And voala, superfast training! :)

**Comment by cklat** [comment]
Did you also managed to find out what the actual bottleneck is hardware-wise for batches that take long to load? Is it the CPU or the RAM? For my setting it seems to be the RAM which is sometimes fully loaded but I'm not sure why this is the case. Sure, like in your case, I have bigger images and more masks in one image but even if I set the batch size to 1 the RAM gets fully loaded sometimes anyway. If I did the math correct, there's no way one of my single images and the amount masks could take up all of my 32 gig memory.

However, the CPU load is also very high for the kernels - depending how much workers you set. But I doubt that the CPU could be the bottleneck in this case. 

Anyway, I will definitely try your workaround and will post back asap.

**Comment by ziyigogogo** [comment]
Actually i have been troubled by this for a while, since i found that when I use "nvidia-smi -l" to monitor my GPU status, it seems that the load of my GPU is super low between batches. 
Will try your suggest and post back.

**Comment by ziyigogogo** [comment]
By the way, which OS you are using? @zungam  
Since https://github.com/matterport/Mask_RCNN/pull/127 shows that on Windows, there is only 1 thread for the generator. I am not sure if this related?

**Comment by benqua** [comment]
@ziyigogogo are you CPU bound?
If this PR did not help you are likely CPU or (disk) IO bound.

**Comment by cklat** [comment]
@zungam I tried your approach but I'm not sure if it really helped to speed up training for my dataset. I believe rather not. 

However, I found another bottleneck if my understanding of the implementation is right. My training setting is basically the same as the nucleus problem from the Data Science Bowl at kaggle. The only difference is that I'm trying to detect mitosis and my images are roughly 4500x4500 pixels and number of instances (masks) in one image might vary between 5-70, thus some images are very sparse regarding the image size and the appearance of the class I'm interested in. 

What I stumbled on, is this piece of code in model.py #1700 that might be the problem for me:

            if not np.any(gt_class_ids > 0):
                continue

If my interpretation of this code snippet and the general training procedure is right, the code is always going to check if the random crop contains the interested class. If that's not the case the cropped image and the original image is skipped and the next image will be taken, getting cropped and eventually again checked if it contains the interested class. 

So, in my case the loading process of the original images and the corresponding masks takes some seconds (depending on the number of masks) and if the code is unlucky in generating the crops that might take a while until the generator got the required number for the batch size.

This might also indicate why some batches take so long to get processed and why the gpu load is very low most of the time.

@waleedka Could you confirm this? And if this is really the case is there a reason why you would only train on 'positive' crops i.e. contain the interesting class and not train on negative samples?

**Comment by ziyigogogo** [comment]
@benqua 
maybe, i could only say, since my cpu usage is not that high either. Did you try this PR and gain a significant improve?
And if so, would please share you hardware condition for reference? Thank you so much

@cklat 
I believe that **waleedka** is currently on his vocation or something since his github activity stopped from the end of april.

**Comment by zungam** [comment]
@cklat 
> Did you also managed to find out what the actual bottleneck is hardware-wise for batches that take long to load?

Sounds weird. I do not experience an increasment of RAM usage. Im sorry i dont know any hardware bottlenecks.

@ziyigogogo 
> which OS you are using?

I use windows. Yes, I believe my implementation only works on windows. But it was a fast work around, and I believe you can google how to make a thread on linux/mac in python to achieve the same. 

@ziyigogogo 
> I tried your workaround and gained about 10% improvement on execute time.

You say you only experience 10% performance gain? Its probably because you are training on a big model such that the GPU exectuion time is allready bigger than CPU execution time. My work around only works if the loading time of batches are very uneven, but faster than GPU execution time on average.

@ziyigogogo 
> Actually i have been troubled by this for a while, since i found that when I use "nvidia-smi -l" to monitor my GPU status, it seems that the load of my GPU is super low between batches."**
This happened to me to before i tried the work around. You are on linux, is the thread working correctly? If it is still superlow between batches it could be that your CPU cant handle all the image processing. How big batch do you use? You should try to decrease the amount of images for each batch. Also, I would suggest adding print statements to the [data_generator](https://github.com/matterport/Mask_RCNN/blob/6c9c82f5feaf5d729a72c67f33e139c4bc71399b/mrcnn/model.py#L1629) and[ load_image_gt](https://github.com/matterport/Mask_RCNN/blob/6c9c82f5feaf5d729a72c67f33e139c4bc71399b/mrcnn/model.py#L1184) to see where the process halts. 
The things that take up most time in the batch loading is usually your own implementation of load_mask.

How man masks do you use for each image? If you have many as I do,  and if you [USE_MINI_MASK](https://github.com/matterport/Mask_RCNN/blob/6c9c82f5feaf5d729a72c67f33e139c4bc71399b/mrcnn/model.py#L1277), it spends time converting all your masks. Also, then these functions will become very slow:
    [bbox = utils.extract_bboxes(mask)](https://github.com/matterport/Mask_RCNN/blob/6c9c82f5feaf5d729a72c67f33e139c4bc71399b/mrcnn/model.py#L1267)
[build_rpn_targets](            rpn_match, rpn_bbox = build_rpn_targets(image.shape, anchors,)

**Comment by zungam** [comment]
@cklat 
> roughly 4500x4500 pixels

This is a huge image. Have you set IMAGE_MIN_DIM = 4500 and/or   IMAGE_MAX_DIM = 4500? In that case you are training on a very large network which should be very slow. If you havent, it means you are downscaling and downscaling all the time to feed your images and masks into your image. You should check if downscaling eats your CPU time. If you are using "crop", then I dont know, but it could be that this also takes up time. To increase training time, I preprossed all images and masks to the correct scale and used "square", so it skipps all scaling and cropping operations

**Comment by cklat** [comment]
@zungam 

Yes, as I said, I upscale the original images then cropping 512x512 images out of the upscaled images. Thus, is set IMAGE_MIN_DIM = 512, IMAGE_MAX_DIM = 512 and IMAGE_RESIZE_MODE = "crop". So everything is basically the same as in the nucleus show case here in the git. 

I set a pdb.set_trace() to sort out the problem. Loading the image and the masks doesn't take that much time, relatively seen. Depending on the number of masks, it might take up to roughly 10 seconds. And besides that, I don't have that much images, in total 27 images, I think.

But doing all the preprocessing might be a good idea, so I  generate the crops all by myself beforehand while ensuring that every crop contains a positive class in the binary mask. Still, I believe if you are working on high resolution images and take relatively small crops from these images, this loop that I mentioned is a huge bottleneck. The only adhoc solution would be to increase the crop size which results in a smaller batch size if you don't have the gpu ram for that. So that would be a trade-off for which I don't know the best practice, i.e. smaller crops/higher batch size vs. bigger crops/smaller batch size
